garrysmod/lua/includes/extensions/vector.lua

```diff
@@ -3,10 +3,262 @@ local meta = FindMetaTable( "Vector" )
+function meta:Nudge(num, vec)
+	self:Add(vec:GetMul(num))
+end
+
+--[[---------------------------------------------------------
+Returns a copy of the nagated vector
+-----------------------------------------------------------]]
+function meta:GetNudge(...)
+	local v = Vector(self)
+	v:Nudge(...)
+	return v
+end
 
+--[[---------------------------------------------------------
+Convets the vector to resized version of itself
+-----------------------------------------------------------]]
+function meta:Resize(num)
+	self:Normalize()
+	self:Mul(num)
+end
+
+--[[---------------------------------------------------------
+Returns a copy of the resized vector
+-----------------------------------------------------------]]
+function meta:GetResized(...)
+	local v = Vector(self)
+	v:Resize(...)
+	return v
+end
+
+--[[---------------------------------------------------------
+Convets the vector to resized midpoint of itself
+-----------------------------------------------------------]]
+function meta:Midpoint(vec)
+	self:Add(vec)
+	self:Mul(0.5)
+end
+
+--[[---------------------------------------------------------
+Returns a copy of the midpoint vector
+-----------------------------------------------------------]]
+function meta:GetMidpoint(...)
+	local v = Vector(self)
+	v:Midpoint(...)
+	return v
+end
+
+--[[---------------------------------------------------------
+Returns a copy of the zero point vector
+-----------------------------------------------------------]]
+function meta:GetZero(...)
+	local v = Vector(self)
+	v:Zero(...)
+	return v
+end
+
+--[[---------------------------------------------------------
+Applies new basis to a vector by the three axises of an angle
+-----------------------------------------------------------]]
+function meta:Basis(ang)
+	local x = self:Dot(ang:Forward())
+	local y = self:Dot(ang:Right())
+	local z = self:Dot(ang:Up())
+	self:SetUnpacked(x, y, z)
+end
+
+--[[---------------------------------------------------------
+Applies new basis to a vector by the three axises of an angle
+-----------------------------------------------------------]]
+function meta:GetBasis(...)
+	local v = Vector(self)
+	v:Basis(...)
+	return v
+end
+
+--[[---------------------------------------------------------
+Projects the vector in to an axis defined by the second vector
+-----------------------------------------------------------]]
+function meta:Project(vec)
+	local vn = vec:GetNormalized()
+	local nm = self:Dot(vn)
+	self:Set(vn)
+	self:Mul(nm)
+end
+
+--[[---------------------------------------------------------
+Projects the vector in to an axis defined by the second vector
+-----------------------------------------------------------]]
+function meta:GetProject(...)
+	local v = Vector(self)
+	v:Project(...)
+	return v
+end
+
+--[[---------------------------------------------------------
+Retrieves the angle between two vectors from 0 to math.pi
+-----------------------------------------------------------]]
+function GetAngle(vec)

```

Comment:
Missing ``meta:`` in front. This should probably renamed to AngleBetween to prevent confusion with vec:Angle(). Also, this could be more efficiently implemented as

```Lua
return math.acos(self:Dot(vec) / math.sqrt(self:LengthSqr() * vec:LengthSqr()))
```

using the formula ``cos(a) = dot(x, y)/(|x|*|y|)``. While you're at it, you should also add an additional optional argument to specify a normal vector representing the plane the vectors lie in. This can be written as

```Lua
return math.atan2(self:Determinant(vec, normal:GetNormalized()), self:Dot(vec))
```

using the formula ``tan(a) = determinant(x, y, n) / dot(x, y)``. Most library functions in GMod don't use type checking so you can get away with

```Lua
function meta:AngleBetween(vec, n)
	if (n == nil) then
		return ...
	end

	return ...
end
```

Reply:
That sounds neat. Thank you !

Reply:
Never noticed the second parameter but it sounds great ;)

Reply:
Done !

```
local a = Vector(3,3)
local b = Vector(3,0)
local c = Vector(0,0,6)
local d = a:GetProject(b)
print(d, a:AngleBetween(b,c))
```

Reply:
VEC{3,0,0}	-0.78539816339745